src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1908,6 +1908,13 @@ abstract class RefChecks extends Transform {
         && !treeInfo.hasExplicitUnit(t)           // suppressed by explicit expr: Unit
         && !isJavaApplication(t)                  // Java methods are inherently side-effecting
       )
+      def checkDiscardValue(t: Tree): Boolean =
+        t.hasAttachment[DiscardedValue.type] && {
+          t.removeAttachment[DiscardedValue.type]
```

Comment:
I think I was avoiding Option. This corner of the API is bothersome.

Reply:
Actually, retronym had already tired of the API, so I will use his improvement. `getAndRemove` isn't very efficient, aside from allocating. It would be nice to have `remove(element): Boolean` with just an indicator that it was removed. I did not touch the calls to update, which remove before adding (presumably to ensure only one attachment of a given type).